### ISSUE TITLE
Improve how TLS enrollment retries on failure

### ISSUE THREAD

**Issue by Smjert** [osquery/osquery#7082]
# Blueprint

## What and How

I'd like to remove the `tls_enroll_max_attempts` flag to be substituted with `tls_enroll_max_backoff_interval` (name is a WIP), and have the enrollment always indefinitely retry on failure, with a quadratic backoff up to the new flag max interval (default 5 minutes).

## Why

1. Users would like to be able to have osquery always infinitely retry enrollment, instead of having to rely sometimes on the service launcher to restart osquery, if configured to do so.

2. There's an unexpected behavior on the amount of enrollment attempts and their patterns when enrollment keeps failing, depending on when it happens. This is due to the multiple places that require osquery to be enrolled and will try to enroll it if it isn't.

3. I don't believe the flag `tls_enroll_max_attempts` is correctly addressing the need of backing off attempts to enroll against an endpoint.

## Details

There are currently 2 main cases where enrollment happens:
1. On startup: osquery will do enrollment attempts up to 2 times the value of `tls_enroll_max_attempts`, with a quadratic backoff that is resetted between the 2 rounds of attempts, which are also done immediately one after the other.
This is due to the fact that the TLSLogForwarder constructor and then TLSLoggerPlugin::setUp method end up requesting osquery to enroll.
After these attempts, osquery exits.

2. When a TLS request fails due to an invalid node key answer from the backend server: re-enrollment is attempted, again a number of times equal to the `tls_enroll_max_attempts`.
But this time the rounds of attempts are infinite, because the plugin that makes the TLS request (either a logger plugin or a config plugin) has its own infinite retry procedure with its own retry interval.

So this means that in one case, on startup osquery exits, while during re-enrollment it potentially does infinite attempts that use two different intervals, one of which changes depending on which plugin did the TLS request that failed.
I would also point out that osquery exiting on startup is not strictly connected to the end of the enrollment attempts,
but it's because a plugin failed to be properly initialized.

In general, osquery not being enrolled is quite blocking and it would make more sense to have osquery
always indefinitely retry enrollment, with a predictable pattern of attempts.

## Alternative implementation

Since one of the main interests of this change is to provide a way to predictably do infinite enrollment retries, it's possible to do a different change and repurpose the `tls_enroll_max_attempts` **0** value to mean infinite retries.
Using that value now, will have osquery enter the enrollment logic but not do any attempts; keep in mind that we already have a flag to disable enrollment (`disable_enrollment`).

This alternative is less of a breaking change, but the backoff behavior would remain the same and so having a configurable high limit for the interval is still very desirable.

Moreover, it would be still desirable to fix the bug of having more attempts happening than requested; ideally this could
be solved by having a persistent pool of retries equal to `tls_enroll_max_attempts`, connected to the enrollment plugin,
so that when it's emptied, then other attempts won't be done.
This would also need a special case handling for the re-enrollment though, to keep the current behavior,
by having the pool resetted when a new TLS request is done.

I believe though that this alternative will make things a bit more complicated and will not fully solve the confusion on the `tls_enroll_max_attempts` behavior.

**Comment by theopolis** [comment]
I like the idea of having `--tls_enroll_max_attempts=0` mean infinite. We can also change the default so that moving forward, with the new max backoff flag, the value of this flag is `=0`.

> it would be still desirable to fix the bug of having more attempts happening than requested

How important is it to people that osquery _only_ attempt the configured number of times? My assumption is this is not very important. I can see this behavior being up for interpretation too. Does max attempts mean stop everything when this limit is reached, or max attempts within a reasonable interval?

The problem that subsequent retries attempts to solve is unstable or lack of connectivity. So I think we can leave some of the behavior up for interpretation, meaning we should not invest a lot of time refactoring code to make sure osquery _never_ goes beyond the max.

If we are _now_ (due to the recent bug fixes) sure enrollment is not attempted concurrently, then a last enrollment attempt timestamp might be sufficient. This way if multiple code paths request enrollment, the first will try, and if that fails the subsequent will observe the timestamp and skip.

**Comment by Smjert** [comment]
Hey @theopolis, thanks for giving this a look!

> How important is it to people that osquery only attempt the configured number of times? My assumption is this is not very important. I can see this behavior being up for interpretation too. Does max attempts mean stop everything when this limit is reached, or max attempts within a reasonable interval?

I think the concern here is having a predictable behavior/pattern of retries; and as you say, having the mixed behavior on startup vs re-enrollment felt a bit strange.
Although admittedly, one of the useful side-effects of osquery exiting, and that we discussed during office hours, is that so one is aware something is going wrong.
This is because at startup the logs won't be even written to the filesystem, but just buffered, so there's "no" awareness of what's happening unless accessing the machine directly.
During re-enrollment instead, the other plugins are already initialized, so there's a possibility to try to understand what is going on.
But as I was saying earlier, osquery exiting is just a side effect of not having a plugin properly initialized, not due to those number of attempts being hit.

>The problem that subsequent retries attempts to solve is unstable or lack of connectivity. So I think we can leave some of the behavior up for interpretation, meaning we should not invest a lot of time refactoring code to make sure osquery never goes beyond the max.

Maybe I'm reading this wrong, but I'm totally for subsequent retries, but as I was saying earlier, having predictable retries means that you know how osquery will behave.
I think the major problem comes from the re-enrollment phase actually, because maybe one wants a quick TLS logger interval; if a request fails because the remote deemed the node invalid, then you would have retries quickly done one after the other, because they will be also affected by the logger interval.
Even if one increases the number of attempts, then those attempts will end and the backoff interval will start from scratch.
If you suddenly have thousands of clients trying to enroll, I think you want for them to stay on the highest interval when they reach it, not have sawtooth pattern.

>If we are now (due to the recent bug fixes) sure enrollment is not attempted concurrently, then a last enrollment attempt timestamp might be sufficient. This way if multiple code paths request enrollment, the first will try, and if that fails the subsequent will observe the timestamp and skip.

This is definitely another good way to model it; although I was trying to push to always do infinite retries because the plugins that rely on enrollment will not work without it anyway, so the idea is that they would all be blocked/waiting for the node key to be retrieved.
Instead having some of them yield trying to enroll, then execute some new error logic, to then retry again, depending on the interval they have been set with, wouldn't they be doing unnecessary work?

Although I can see one possible case to keep in mind, which is that there might be a code path/thread which calls getNodeKey but it's bad if gets blocked on it, in the case of enrollment failure, because it also has other valuable work to do.

**Comment by Smjert** [comment]
Another thing I would like to add is that currently each plugin you enable that needs TLS (distributed, config, logger) will also do its number of attempts.
So when enabling the logger plugin, as already said it will try 2*`tls_enroll_max_attempts` times.
If you also enable the config plugin, then `config_tls_max_attempts` * `tls_enroll_max_attempts` attempts will be added to the previous ones, and so on for the other plugins.

**Comment by directionless** [comment]
As @Smjert said, this came up in prior office hours. 

I am somewhat opposed to the proposal, but strongly in favor of the alternate ones. That is:
1. I do not think we should default to retry-forever. I think broken things should exist
1. I like `tls_enroll_max_attempts=0` meaning infinite. This feels like it has good prior art. (`-1` would also work, if we really want to support `0`)
1. I'd support a default higher than `3`. Whatever 15-30 minutes works out to, maybe?
1. Fixing bugs about duplicate enrollments seems reasonable, but like Teddy I would not have thought it was a critical issue

As you call out, I think there's a lot of confusion because this blurs the initial enrollment with a subsequent re-enrollment. I'm not sure there's a simple answer there. They feel fundamentally different. No initial enrollment _generally_ means no configuration. There is little point in proceeding. But a failure to re-enroll has viable fallbacks. 

So maybe a related question here is what should happen when a server requests re-enrollment. Should we drop all local state, and consider a new client? Or treat existing state as something to preserve

**Comment by Smjert** [comment]
I think we can tackle this in steps then and first implement the infinite retry for when the value is set to 0 and I will also add a max interval flag to control the interval between retries, because otherwise the time increases indefinitely.

The max interval is also useful in the non-infinite case, because having high intervals means osquery taking more time than needed to enroll (or enroll again if it's re-enrollment).

The duplicate enrollment issue is not critical, but I think there's value to have flags that behave in a single specific way vs changing behavior depending on how osquery is run and how unrelated code evolves.
It's easier to debug, easier to explain, easier to understand.

**Comment by directionless** [comment]
Yeah, in steps seems reasonable, and your proposed infinite retry for when the value is set to 0 and max interval flag both seem non-controversial and great to me.
